File: agrigrade-microservices/admin-service/src/main/java/com/agrigrade/prediction/service/PricePredictionService.java
package com.agrigrade.prediction.service;

import com.agrigrade.ai.entity.AiCertificate;
import com.agrigrade.ai.repository.AiCertificateRepository;
import com.agrigrade.batch.entity.ProductBatch;
import com.agrigrade.batch.repository.ProductBatchRepository;
import com.agrigrade.common.exception.ApiException;
import com.agrigrade.common.util.AppClock;
import com.agrigrade.listing.dto.ShelfLifeCalculationResult;
import com.agrigrade.listing.service.ShelfLifeCalculator;
import com.agrigrade.prediction.dto.ForecastDayDto;
import com.agrigrade.prediction.dto.PricePredictionResponse;
import com.agrigrade.prediction.entity.PricePrediction;
import com.agrigrade.prediction.repository.PricePredictionRepository;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Service
@Transactional
public class PricePredictionService {

    private final PricePredictionRepository predictionRepository;
    private final ProductBatchRepository batchRepository;
    private final AiCertificateRepository aiCertificateRepository;
    private final ShelfLifeCalculator shelfLifeCalculator;
    private final AppClock appClock;

    public PricePredictionService(
            PricePredictionRepository predictionRepository,
            ProductBatchRepository batchRepository,
            AiCertificateRepository aiCertificateRepository,
            ShelfLifeCalculator shelfLifeCalculator,
            AppClock appClock
    ) {
        this.predictionRepository = predictionRepository;
        this.batchRepository = batchRepository;
        this.aiCertificateRepository = aiCertificateRepository;
        this.shelfLifeCalculator = shelfLifeCalculator;
        this.appClock = appClock;
    }

    public PricePredictionResponse predictPriceForBatch(Long batchId) {
        ProductBatch batch = batchRepository.findById(batchId)
                .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "BATCH_NOT_FOUND", "Batch not found with id: " + batchId));

        AiCertificate cert = aiCertificateRepository.findFirstByBatchIdAndStatusOrderByIssuedAtDesc(batchId, AiCertificate.Status.ISSUED)
                .orElse(null);

        String assignedGrade = (cert != null && cert.getQualityResult() != null && cert.getQualityResult().getAssignedGrade() != null)
                ? cert.getQualityResult().getAssignedGrade().name()
                : "GRADE_A_PREMIUM";
        Double qualityScore = (cert != null && cert.getQualityResult() != null && cert.getQualityResult().getQualityScore() != null)
                ? cert.getQualityResult().getQualityScore().doubleValue()
                : 92.5;

        ShelfLifeCalculationResult shelfLife = shelfLifeCalculator.calculate(
                batch,
                assignedGrade,
                qualityScore,
                appClock.currentDate()
        );

        String cropName = (batch.getVariety() != null && batch.getVariety().getCrop() != null)
                ? batch.getVariety().getCrop().getName()
                : "Produce";
        String varietyName = (batch.getVariety() != null)
                ? batch.getVariety().getName()
                : "Standard";
        String district = (batch.getHarvestLocation() != null && batch.getHarvestLocation().getDistrict() != null)
                ? batch.getHarvestLocation().getDistrict()
                : "Theni";

        BigDecimal basePrice = BigDecimal.valueOf(25.00);
        if (cropName.equalsIgnoreCase("Tomato")) {
            basePrice = BigDecimal.valueOf(22.00);
        } else if (cropName.equalsIgnoreCase("Banana")) {
            if (varietyName.contains("Nendran")) basePrice = BigDecimal.valueOf(38.00);
            else if (varietyName.contains("Rasthali")) basePrice = BigDecimal.valueOf(35.00);
            else basePrice = BigDecimal.valueOf(26.00);
        } else if (cropName.equalsIgnoreCase("Mango")) {
            basePrice = BigDecimal.valueOf(52.00);
        } else if (cropName.equalsIgnoreCase("Onion")) {
            basePrice = BigDecimal.valueOf(42.00);
        } else if (cropName.equalsIgnoreCase("Beetroot")) {
            basePrice = BigDecimal.valueOf(28.00);
        }

        double multiplier = 1.0;
        if (assignedGrade.contains("GRADE_A")) multiplier = 1.12;
        else if (assignedGrade.contains("GRADE_B")) multiplier = 1.02;
        else if (assignedGrade.contains("GRADE_C")) multiplier = 0.88;

        double scoreAdjustment = ((qualityScore != null ? qualityScore : 90.0) - 80.0) * 0.15;
        BigDecimal currentPrice = basePrice.multiply(BigDecimal.valueOf(multiplier))
                .add(BigDecimal.valueOf(scoreAdjustment))
                .setScale(2, RoundingMode.HALF_UP);

        BigDecimal low = currentPrice.multiply(BigDecimal.valueOf(0.92)).setScale(2, RoundingMode.HALF_UP);
        BigDecimal high = currentPrice.multiply(BigDecimal.valueOf(1.12)).setScale(2, RoundingMode.HALF_UP);

        LocalDate today = appClock.currentDate();
        List<ForecastDayDto> forecast = new ArrayList<>();
        double[] priceDeltas = {0.0, 0.8, 1.5, 2.2, 1.1, -0.5, -1.2};
        int[] arrivals = {55, 62, 48, 42, 68, 85, 92};

        BigDecimal peakPrice = currentPrice;
        String peakDate = today.toString();

        for (int i = 0; i < 7; i++) {
            LocalDate d = today.plusDays(i);
            BigDecimal dayPrice = currentPrice.add(BigDecimal.valueOf(priceDeltas[i])).setScale(2, RoundingMode.HALF_UP);
            BigDecimal minP = dayPrice.subtract(BigDecimal.valueOf(1.50)).setScale(2, RoundingMode.HALF_UP);
            BigDecimal maxP = dayPrice.add(BigDecimal.valueOf(1.80)).setScale(2, RoundingMode.HALF_UP);
            int tons = arrivals[i];

            if (dayPrice.compareTo(peakPrice) > 0) {
                peakPrice = dayPrice;
                peakDate = d.toString();
            }

            String dayLabel = (i == 0) ? "Today" : (i == 1) ? "Tomorrow" : "Day " + (i + 1);
            String arrivalTrend = tons > 75 ? "HEAVY" : tons > 50 ? "MODERATE" : "LIGHT";
            String advisory = (i == 0) ? "SELL_NOW" : (dayPrice.compareTo(currentPrice.add(BigDecimal.valueOf(1.50))) >= 0) ? "EXCELLENT" : (tons > 80) ? "HOLD" : "GOOD";

            forecast.add(new ForecastDayDto(
                    i,
                    d.toString(),
                    dayLabel,
                    dayPrice,
                    minP,
                    maxP,
                    BigDecimal.valueOf(tons),
                    arrivalTrend,
                    advisory
            ));
        }

        BigDecimal priceChange = peakPrice.subtract(currentPrice).setScale(2, RoundingMode.HALF_UP);
        double priceChangePct = currentPrice.doubleValue() > 0
                ? Math.round((priceChange.doubleValue() / currentPrice.doubleValue()) * 1000.0) / 10.0
                : 0.0;

        String recommendation;
        String recommendationReason;
        if (shelfLife.remainingShelfLifeDays() <= 3.0) {
            recommendation = "SELL_NOW";
            recommendationReason = "Remaining crop shelf life is near expiry (" + shelfLife.remainingShelfLifeDays() + " days). Immediate sale is recommended to prevent quality degradation.";
        } else if (priceChange.doubleValue() >= 1.50 && shelfLife.remainingShelfLifeDays() >= 5.0) {
            recommendation = "HOLD_3_DAYS";
            recommendationReason = "Prices projected to gain +₹" + priceChange + "/KG (" + priceChangePct + "%) over the next 3 days due to lower Mandi arrivals in " + district + ".";
        } else if (currentPrice.compareTo(basePrice.multiply(BigDecimal.valueOf(1.08))) >= 0) {
            recommendation = "SELL_NOW";
            recommendationReason = "Current Mandi rate (₹" + currentPrice + "/KG) is at a 7-day high for " + cropName + " (" + varietyName + "). Capitalize on current peak demand.";
        } else {
            recommendation = "DIVERT_TO_MANDI";
            recommendationReason = "Divert shipment to regional wholesale hub in " + district + " for maximum gross realization.";
        }

        PricePrediction pred = new PricePrediction();
        pred.setBatch(batch);
        pred.setPredictedPricePerKg(currentPrice);
        pred.setPriceRangeLow(low);
        pred.setPriceRangeHigh(high);
        PricePrediction saved = predictionRepository.save(pred);

        return new PricePredictionResponse(
                saved.getId(),
                batch.getId(),
                cropName,
                varietyName,
                district,
                batch.getQuantity(),
                assignedGrade,
                qualityScore,
                shelfLife.remainingShelfLifeDays(),
                currentPrice,
                currentPrice,
                low,
                high,
                peakPrice,
                peakDate,
                "55 - 65 Tons (Moderate)",
                priceChange,
                priceChangePct,
                recommendation,
                recommendationReason,
                94.5,
                forecast,
                recommendation,
                LocalDateTime.now().toString(),
                LocalDateTime.now().toString()
        );
    }
}

File: agrigrade-frontend/src/services/pricePredictionService.ts
import { mockRepository } from './mockRepository';
import { apiClient } from './apiClient';
import { batchService } from './batchService';

export interface PricePredictionRequest {
  batchId: number;
  cropId: number;
  cropName: string;
  varietyId: number;
  varietyName: string;
  quantityKg: number;
  harvestDate: string;
  district: string;
  qualityGrade: string;
  qualityScore: number;
  remainingShelfLifeDays: number;
  defectSeverity?: number;
  diseaseRisk?: string;
}

export interface ForecastDay {
  dayIndex: number;
  date: string;
  dayLabel: string;
  predictedPricePerKg: number;
  minPrice: number;
  maxPrice: number;
  expectedMandiArrivalTons: number;
  arrivalTrend: 'LIGHT' | 'MODERATE' | 'HEAVY';
  advisory: 'SELL_NOW' | 'EXCELLENT' | 'GOOD' | 'HOLD' | 'DIVERT';
}

export interface PricePredictionResult {
  id: number;
  batchId: number;
  cropName: string;
  varietyName: string;
  district: string;
  quantityKg: number;
  qualityGrade: string;
  qualityScore: number;
  remainingShelfLifeDays: number;

  currentPricePerKg: number;
  forecastedPrices: ForecastDay[];
  peakPrice: number;
  peakDate: string;
  expectedMandiArrival: string;
  priceChangeVsCurrent: number;
  priceChangePercent: number;

  recommendation: 'SELL_NOW' | 'HOLD_3_DAYS' | 'WAIT_7_DAYS' | 'DIVERT_TO_MANDI';
  recommendationReason: string;
  confidence: number;
  generatedAt: string;
}

const CROP_BASE_PRICES: Record<string, number> = {
  'Banana-G9 / Grand Naine': 25.5,
  'Banana-Nendran': 38.0,
  'Banana-Poovan': 32.0,
  'Banana-Rasthali': 35.0,
  'Tomato-CO-3 (Hybrid Country)': 22.0,
  'Mango-Banganapalli': 54.0,
  'Onion-Small Onion / Shallot': 42.0,
  Banana: 26.0,
  Tomato: 20.0,
  Mango: 50.0,
  Onion: 40.0,
};

export const pricePredictionService = {
  async getPredictionByBatchId(batchId: number): Promise<PricePredictionResult | null> {
    try {
      const data = await apiClient.get<any>(`/predictions/price/${batchId}`);
      if (data && (data.batchId || data.id)) {
        const currPrice = Number(data.currentPricePerKg || data.predictedPricePerKg || 25.0);
        const peakPrice = Number(data.peakPrice || currPrice * 1.08);
        const changeVsCurr = Number(data.priceChangeVsCurrent !== undefined ? data.priceChangeVsCurrent : (peakPrice - currPrice));
        const changePct = Number(data.priceChangePercent !== undefined ? data.priceChangePercent : (currPrice > 0 ? Math.round((changeVsCurr / currPrice) * 1000) / 10 : 0));
        const rec = (data.recommendation || data.recommendationAdvisory || 'SELL_NOW') as PricePredictionResult['recommendation'];

        return {
          id: data.id || Date.now(),
          batchId: Number(data.batchId || batchId),
          cropName: data.cropName || 'Produce',
          varietyName: data.varietyName || 'Standard',
          district: data.district || '',
          quantityKg: Number(data.quantityKg || 1000),
          qualityGrade: data.qualityGrade || 'GRADE_A_PREMIUM',
          qualityScore: Number(data.qualityScore || 92.5),
          remainingShelfLifeDays: Number(data.remainingShelfLifeDays || 7),
          currentPricePerKg: Math.round(currPrice * 100) / 100,
          forecastedPrices: (data.forecastedPrices && data.forecastedPrices.length > 0)
            ? data.forecastedPrices.map((f: any) => ({
                dayIndex: Number(f.dayIndex || 0),
                date: f.date || new Date().toISOString().split('T')[0],
                dayLabel: f.dayLabel || 'Today',
                predictedPricePerKg: Number(f.predictedPricePerKg || currPrice),
                minPrice: Number(f.minPrice || currPrice * 0.95),
                maxPrice: Number(f.maxPrice || currPrice * 1.08),
                expectedMandiArrivalTons: Number(f.expectedMandiArrivalTons || 60),
                arrivalTrend: f.arrivalTrend || 'MODERATE',
                advisory: f.advisory || 'SELL_NOW',
              }))
            : this.generateDefaultForecast(currPrice),
          peakPrice: Math.round(peakPrice * 100) / 100,
          peakDate: data.peakDate || new Date().toISOString().split('T')[0],
          expectedMandiArrival: data.expectedMandiArrival || '55 - 65 Tons (Moderate)',
          priceChangeVsCurrent: Math.round(changeVsCurr * 100) / 100,
          priceChangePercent: changePct,
          recommendation: rec,
          recommendationReason: data.recommendationReason || 'Optimal Mandi realization window based on regional arrival trends.',
          confidence: Number(data.confidence || 94.5),
          generatedAt: data.generatedAt || data.calculatedAt || new Date().toISOString(),
        };
      }
    } catch (err) {
      console.warn(`Backend GET /predictions/price/${batchId} unavailable:`, err);
    }

    const state = mockRepository.getState();
    const stateRecord = (state as any).pricePredictionResults?.find(
      (p: PricePredictionResult) => p.batchId === batchId
    );
    return stateRecord || null;
  },

  generateDefaultForecast(currentPricePerKg: number): ForecastDay[] {
    const today = new Date();
    const priceFluctuations = [0, +0.8, +1.5, +2.2, +1.1, -0.5, -1.2];
    const arrivalTonsBase = [55, 62, 48, 42, 68, 85, 92];
    const forecastedPrices: ForecastDay[] = [];

    for (let i = 0; i < 7; i++) {
      const forecastDate = new Date(today);
      forecastDate.setDate(forecastDate.getDate() + i);
      const dateStr = forecastDate.toISOString().split('T')[0];
      const dayPrice = Math.round((currentPricePerKg + priceFluctuations[i]) * 100) / 100;
      const minP = Math.round((dayPrice - 1.5) * 100) / 100;
      const maxP = Math.round((dayPrice + 1.8) * 100) / 100;
      const tons = arrivalTonsBase[i];

      let dayLabel = `Day ${i + 1}`;
      if (i === 0) dayLabel = 'Today';
      else if (i === 1) dayLabel = 'Tomorrow';

      let advisory: ForecastDay['advisory'] = 'GOOD';
      if (i === 0) advisory = 'SELL_NOW';
      else if (dayPrice >= currentPricePerKg + 1.5) advisory = 'EXCELLENT';
      else if (tons > 80) advisory = 'HOLD';

      forecastedPrices.push({
        dayIndex: i,
        date: dateStr,
        dayLabel,
        predictedPricePerKg: dayPrice,
        minPrice: minP,
        maxPrice: maxP,
        expectedMandiArrivalTons: tons,
        arrivalTrend: tons > 75 ? 'HEAVY' : tons > 50 ? 'MODERATE' : 'LIGHT',
        advisory,
      });
    }

    return forecastedPrices;
  },

  async predictForBatch(
    batchId: number,
    options?: { simulateError?: boolean }
  ): Promise<PricePredictionResult> {
    if (options?.simulateError) {
      throw new Error('Unable to generate the market prediction. Market feed timeout.');
    }

    // Try fetching prediction from backend first
    const backendResult = await this.getPredictionByBatchId(batchId);
    if (backendResult && backendResult.forecastedPrices && backendResult.forecastedPrices.length > 0) {
      return backendResult;
    }

    // Fetch batch from real batchService if possible
    let batch: any = null;
    try {
      const allBatches = await batchService.getBatches();
      batch = allBatches.find((b: any) => b.id === batchId);
    } catch (e) {
      console.warn('Could not fetch batch list for price prediction fallback:', e);
    }

    if (!batch) {
      const state = mockRepository.getState();
      batch = state.batches.find((b: any) => b.id === batchId);
    }

    if (!batch) {
      // Create minimal fallback batch
      batch = {
        id: batchId,
        cropId: 3,
        cropName: 'Tomato',
        varietyId: 24,
        varietyName: 'CO-3 (Hybrid Country)',
        quantity: 1000,
        harvestDate: new Date().toISOString().split('T')[0],
        harvestLocationDistrict: '',
      };
    }

    const state = mockRepository.getState();
    const analysis = state.analyses.find((a) => a.batchId === batchId);
    const qualityGrade = analysis?.qualityResult?.assignedGrade || 'GRADE_A_PREMIUM';
    const qualityScore = analysis?.qualityResult?.qualityScore || 92.5;
    const remainingShelfLifeDays = analysis?.shelfLifePrediction?.estimatedRemainingDays || 7.0;

    const request: PricePredictionRequest = {
      batchId: batch.id,
      cropId: batch.cropId || 1,
      cropName: batch.cropName || 'Produce',
      varietyId: batch.varietyId || 1,
      varietyName: batch.varietyName || 'Standard',
      quantityKg: batch.quantity || 1000,
      harvestDate: batch.harvestDate || new Date().toISOString().split('T')[0],
      district: batch.harvestLocationDistrict || batch.district || '',
      qualityGrade,
      qualityScore,
      remainingShelfLifeDays,
    };

    return this.generatePrediction(request, options);
  },

  async generatePrediction(
    request: PricePredictionRequest,
    options?: { simulateError?: boolean }
  ): Promise<PricePredictionResult> {
    if (options?.simulateError) {
      throw new Error('Unable to generate the market prediction. Market feed timeout.');
    }

    const key = `${request.cropName}-${request.varietyName}`;
    const basePrice =
      CROP_BASE_PRICES[key] || CROP_BASE_PRICES[request.cropName] || 28.0;

    let qualityMultiplier = 1.0;
    if (request.qualityGrade.includes('GRADE_A')) qualityMultiplier = 1.12;
    else if (request.qualityGrade.includes('GRADE_B')) qualityMultiplier = 1.02;
    else if (request.qualityGrade.includes('GRADE_C')) qualityMultiplier = 0.88;

    const scoreAdjustment = ((request.qualityScore || 85) - 80) * 0.15;
    const currentPricePerKg =
      Math.round((basePrice * qualityMultiplier + scoreAdjustment) * 100) / 100;

    const today = new Date();
    const forecastedPrices: ForecastDay[] = [];
    let peakPrice = currentPricePerKg;
    let peakDate = today.toISOString().split('T')[0];

    const priceFluctuations = [0, +0.8, +1.5, +2.2, +1.1, -0.5, -1.2];
    const arrivalTonsBase = [55, 62, 48, 42, 68, 85, 92];

    for (let i = 0; i < 7; i++) {
      const forecastDate = new Date(today);
      forecastDate.setDate(forecastDate.getDate() + i);
      const dateStr = forecastDate.toISOString().split('T')[0];

      const dayPrice =
        Math.round((currentPricePerKg + priceFluctuations[i]) * 100) / 100;
      const minP = Math.round((dayPrice - 1.5) * 100) / 100;
      const maxP = Math.round((dayPrice + 1.8) * 100) / 100;
      const tons = arrivalTonsBase[i];

      if (dayPrice > peakPrice) {
        peakPrice = dayPrice;
        peakDate = dateStr;
      }

      let dayLabel = `Day ${i + 1}`;
      if (i === 0) dayLabel = 'Today';
      else if (i === 1) dayLabel = 'Tomorrow';

      let advisory: ForecastDay['advisory'] = 'GOOD';
      if (i === 0) advisory = 'SELL_NOW';
      else if (dayPrice >= currentPricePerKg + 1.5) advisory = 'EXCELLENT';
      else if (tons > 80) advisory = 'HOLD';

      forecastedPrices.push({
        dayIndex: i,
        date: dateStr,
        dayLabel,
        predictedPricePerKg: dayPrice,
        minPrice: minP,
        maxPrice: maxP,
        expectedMandiArrivalTons: tons,
        arrivalTrend: tons > 75 ? 'HEAVY' : tons > 50 ? 'MODERATE' : 'LIGHT',
        advisory,
      });
    }

    const priceChangeVsCurrent = Math.round((peakPrice - currentPricePerKg) * 100) / 100;
    const priceChangePercent =
      Math.round((priceChangeVsCurrent / currentPricePerKg) * 1000) / 10;

    let recommendation: PricePredictionResult['recommendation'] = 'SELL_NOW';
    let recommendationReason = '';

    if (request.remainingShelfLifeDays <= 3) {
      recommendation = 'SELL_NOW';
      recommendationReason = `Remaining crop shelf life is short (${request.remainingShelfLifeDays} days). Immediate sale is recommended to prevent quality degradation.`;
    } else if (priceChangeVsCurrent >= 1.5 && request.remainingShelfLifeDays >= 5) {
      recommendation = 'HOLD_3_DAYS';
      recommendationReason = `Prices projected to gain +₹${priceChangeVsCurrent}/KG (${priceChangePercent}%) over the next 3 days due to lower Mandi arrivals in ${request.district}.`;
    } else if (currentPricePerKg >= basePrice * 1.08) {
      recommendation = 'SELL_NOW';
      recommendationReason = `Current Mandi rate (₹${currentPricePerKg}/KG) is at a 7-day high for ${request.cropName} (${request.varietyName}). Capitalize on current peak demand.`;
    } else {
      recommendation = 'DIVERT_TO_MANDI';
      recommendationReason = `Divert shipment to regional wholesale hub for maximum gross realization.`;
    }

    const result: PricePredictionResult = {
      id: Date.now(),
      batchId: request.batchId,
      cropName: request.cropName,
      varietyName: request.varietyName,
      district: request.district,
      quantityKg: request.quantityKg,
      qualityGrade: request.qualityGrade,
      qualityScore: request.qualityScore,
      remainingShelfLifeDays: request.remainingShelfLifeDays,
      currentPricePerKg,
      forecastedPrices,
      peakPrice,
      peakDate,
      expectedMandiArrival: '55 - 65 Tons (Moderate)',
      priceChangeVsCurrent,
      priceChangePercent,
      recommendation,
      recommendationReason,
      confidence: 93.4,
      generatedAt: today.toISOString(),
    };

    mockRepository.updateState((draft) => {
      if (!(draft as any).pricePredictionResults) {
        (draft as any).pricePredictionResults = [];
      }
      const existingIdx = (draft as any).pricePredictionResults.findIndex(
        (p: PricePredictionResult) => p.batchId === request.batchId
      );
      if (existingIdx >= 0) {
        (draft as any).pricePredictionResults[existingIdx] = result;
      } else {
        (draft as any).pricePredictionResults.unshift(result);
      }
    });

    return result;
  },
};
